src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -878,9 +878,6 @@ address MacroAssembler::trampoline_call(Address entry, CodeBuffer* cbuf,
     bool in_scratch_emit_size = false;
 #ifdef COMPILER2
     if (check_emit_size) {
-      assert(StubRoutines::aarch64::complete() &&
-             Thread::current()->is_Compiler_thread() &&
-             Compile::current()->output() != NULL, "not in output phase of C2 compilation");
```

Comment:
The code is under `#ifdef COMPILER2` which assumes that it is executed only for C2. Which make it confusing because it CAN be executed by C1.
There is check at line 885 `is_c2_compile()` which is true only for C2. May be we can rearrange code to make it less confusing.
On other hand if `check_emit_size` could be `true` for C1 why we should emit trampoline for it and not for C2?

Reply:
@vnkozlov I will leave it to you to decide whether to sponsor as-is or look for additional changes. The failures seem intermittent and have not re-occurred in the next two CI runs.

Reply:
Leave it out. It does nothing useful; I think it was just a precaution.

Reply:
> The failures seem intermittent and have not re-occurred in the next two CI runs.

@dholmes-ora It is intermittent because we get unlucky and `Compile::current()->output()` either crashes or returns `NULL`.

Reply:
@vnkozlov 
> On other hand if `check_emit_size` could be `true` for C1 why we should emit trampoline for it and not for C2?

IMHO `check_emit_size` should always be `false` for C1 because C1 does not have such functionality. This was not caught by testing.

The problem is what the default value `check_emit_size` should have. The current default value `true` leads to confusion for the cases where 'check emit size' has no meaning like C1 or stub generation. 

A possible solution simplifying everything can be:

```
virtual bool MacroAssembler::emit_trampoline_stub(...) {
  The default implementation which always generates a trampoline code.
}


address MacroAssembler::trampoline_call(Address entry, CodeBuffer* cbuf) {
  bool need_trampoline = far_branches();
  if (!need_trampoline && entry.rspec().type() == relocInfo::runtime_call_type && !CodeCache::contains(entry.target())) {
...
  }

  if (cbuf) cbuf->set_insts_mark();
  relocate(entry.rspec());
  if (!need_trampoline) {
    bl(entry.target());
  } else {
    if (!emit_trampoline_stub(offset(), entry.target())) {
      postcond(pc() == badAddress);
      return NULL; // CodeCache is full
    }
    bl(pc());
  }
  // just need to return a non-null address
  postcond(pc() != badAddress);
  return pc();
}

bool C2_MacroAssembler::emit_trampoline_stub(...) {
   auto phase_output = Compile::current()->output();
   if (phase_output !=NULL && phase_output->in_scratch_emit_size()) {
      // We don't want to emit a trampoline if C2 is generating dummy
      // code during its branch shortening phase.
     return true;
   }
   return MacroAssembler::emit_trampoline_stub(...);
}
```

I can submit PR with it if there is interest in the solution.
What do you think?

Reply:
> > I can submit PR with it if there is interest in the solution.
> > What do you think?
> 
> @eastig Please do that. Your code change suggestion is good.

I've created PR: https://github.com/openjdk/jdk/pull/9782